transcom/mymove
# MB-16104: v2 prime api for new fields

## [Jira ticket](https://dp3.atlassian.net/browse/MB-16104)

## Summary

Initially, this followed the instructions in this [ADR](https://transcom.github.io/mymove-docs/docs/adrs/api-versioning/) and implements the changes to support a singular new API end point `prime/v2/move-task-orders/{moveTaskOrderID}`. I followed these three example PRs that were linked in the ADR:

- [For yaml changes](https://github.com/transcom/mymove/pull/10816/commits/0f387d655e098e589b10ab27b388bc771e873f09)
- [For handler changes](https://github.com/transcom/mymove/pull/10816/commits/0c4be3bab159a49f83e407830a9fecb9d27278da)
- [For service changes](https://github.com/transcom/mymove/compare/0c4be3b...8ad4aed)

If you are trying to get a sense of what making these changes was like, please take a look at the individual commits. 

## My thoughts on this process

### Yaml changes

Making these changes was quite cumbersome. For a larger entity like move_task_order we need a lot of the existing definitions. This meant a lot of manual copy pasting. This also made it difficult to keep the changes in line with main since the original prime.yaml was actively being modified. 

### Handler changes

This was fairly straight-forward yet cumbersome. Adding the model to payload functions for such a large entity meant we were primarily copying functions over with minor modifications. 

### Service changes

Outside of the initial flag setup, this was the most smooth portion of the changes. We had a service for the v1 API and a service for the v2 API. If you diff the two files you can see what the differences are between the two. We could potentially make some enhancements on how we are setting these flags and add some additional logic to make the behavior more consistent between when API calls are being made and when tests are being run. Right now the flag that we are using is nil when we are running tests. (There are some comments about this [[1](https://github.com/transcom/mymove/pull/10952#discussion_r1253262344) | [2](https://github.com/transcom/mymove/pull/10952#discussion_r1253264179)].)

### Tests

The tests are in many ways just duplication of existing tests with minor edits. It's important to call out that the approach that we are taking here make it so that the default behavior utilizes the newer v2 service. This means all existing tests will point to the new v2 service.

## Deviations from the suggested approach

### Yaml

I ended up moving some of the existing definitions into their own files so that they can be used across yaml files. These changes are done in #11042. Currently this PR is set to merge into that branch so that we could better isolate these different changes. This will "theoretically" cut down on the duplicated code and make it easier to maintain. Imo, it also makes it a bit easier to identify what structures have changed from v1 to v2. For example in this case its the MoveTaskOrder and Order objects. For viewing purposes a developer can still reference the generated swagger files or they can use [go-swagger to expand/flatten](https://goswagger.io/use/transform.html) to see all of the definitions pulled into a single file. 

### Services/Flags

Previously in order to make the service changes we had introduced some additional flags in the appContext to determine the version of the API that was being called. This obfuscated complexity by making the flag setup happen "behind the scenes" and was something that had impacts on how tests were running without being very explicit. I ended up removing the flags and going with a service approach that was similar to [option 1 ](https://github.com/transcom/mymove/pull/10816/commits/0913f2f4625450a2ac43b358e1890d17d77b747d)where we would just introduce a new service as needed. For this specific instance, I ended up just overriding the existing services as the change was fairly minimal and didn't have significant downsides. If we were introducing say 5 new eager loaded fields that would not always be needed we could have created a separate function to do this and created some reusable local functions that abstracted away some of the shared logic. 

## Other Callouts

During this implementation there were a handful of changes to the model to payload file. Since we had to basically duplicate and adapt those existing functions for the v2 generated files this means that this is an area where there might be errors due to manual input. When reviewing this PR pay special attention to the model to payload functions to ensure that the new version matches the existing versions for all fields that haven't changed.

We do expect this approach to evolve and adapt over time. The purpose of this change is to introduce some plumbing and foundation for future changes so that folks have an example they can refer to if they need direction on how to structure their changes.

If this is merged, we are planning to update the ADR to reflect the changes made here and to potentially add some additional documentation documenting the findings we discovered during this process.

### Considerations for the future

We may want to add some codeowners to the generated swagger files so that there is a manual process for accepting changes that may impact APIs.

We also probably want to spend some time spreading awareness around how modifying existing services can and may impact existing APIs and what the implications of those changes are when we have consumers of our APIs.

It may be worthwhile to explore some other libraries or writing an external wrapper that enables us to prevent the duplication of code around the generated files and the model to payload conversions when the objects are identical.

## Verification Steps for the Reviewer

Take a look at the `Other Callouts` section.

1. Verify the new API is working correctly
2. Verify the old API is still working as expected
3. Verify the new prime_v2.yaml file and corresponding docs
4. Verify the existing prime.yaml file and corresponding docs

### Setup to Run the Code

- [Instructions for starting storybook](https://transcom.github.io/mymove-docs/docs/frontend/setup/run-storybook)
- [Instructions for starting the MilMove application](https://transcom.github.io/mymove-docs/docs/about/application-setup/milmove-local-client/)
- [Instructions for running tests](https://transcom.github.io/mymove-docs/docs/about/development/testing)

### How to test

For folks that have made some changes to the payload to model file recently, if you can test scenarios that have those newly added/modified fields populated that would be ideal.

1. Compare and contrast the responses between the v1 and v2 API where :moveTaskOrderID is a path variable
    a. In postman make a get call to the v1 API `https://primelocal:9443/prime/v1/move-task-orders/:moveTaskOrderID`
    b. In postman make a get call to the v2 API `https://primelocal:9443/prime/v2/move-task-orders/:moveTaskOrderID`
    
    You should see they are identical except in the v2 version the fields `methodOfPayment`, `naics`, `packingAndShippingInstructions`,  and `supplyAndServicesCostEstimate` are populated within the orders object and the `contractNumber` is populated within the top level object.
    
    ![image](https://github.com/transcom/mymove/assets/84742904/1848c767-33c4-4cb7-8b61-4db8ccfcb5b7)

2. You can also compare the v1 API response on this branch with main to ensure that the v1 branch has not changed.
3. To test the docs locally you can close the backend server since it runs on port 8080 and then run `openapi preview-docs swagger/prime_v2.yaml`. You can then visit `localhost:8080` and you should be able compare and contrast the v1 API docs with the v2 docs. Ensure that the new fields exist on the v2 doc. 
4. If you want to ensure that the v1 API hasn't changed you can ensure that the prime.yaml in the swagger folder hasn't been modified.
